### ISSUE TITLE
SFTP Polling

### ISSUE THREAD

**Issue by scleary1cs** [CDCgov/trusted-intermediary#1080]
# Story
As a CA, so that my files are picked up in a timely manner and I don't lock out users, I need the SFTP service to check for new files.

## Pre-conditions
- [x] #1077 is complete (#1077 is complete enough as of 07/10/2024).

## Acceptance Criteria
- [x] New files are checked for on a configured basis.
- [x] When a new file is discovered, #1077 work will kick in.

### Research
- [ ] How often should we be checking for new files
- [ ] When (what time window) should we be checking for new files

###  Engineering
- [x] POC azure function on a cron? Need to make sure this wouldn't scale up and create too many tasks
  - [x] Add terraform for Function 
  - [x] Add github actions to deploy function
  - [x] Add azure function that sends a queue message based on a CRON schedule
- [x] Add queue reader that reads from queue and then kicks off connecting to external SFTP service
  - [x] (including refactoring existing reader where appropriate)
  - [x] Add tests for this reader 
- [x] Update CRON details based on info from CA
- [x] Set visibility timeout to a high enough value, remove the TODO for this (Azure default is 30 seconds)
 `	// If not specified, the default value is 0. Specifies the new visibility timeout value,
	// in seconds, relative to server time. The value must be larger than or equal to 0, and cannot be
	// larger than 7 days. The visibility timeout of a message cannot be
	// set to a value later than the expiry time. VisibilityTimeout
	// should be set to a value smaller than the time-to-live value.`
- [x] Set message expiration for all Queues.
- [x] Improve copyFiles error so that the error is actionable (do we give up/delete queue message or do we keep trying?) - I think this is waiting on discussion with @halprin and maybe some others?
  - [x] After group discussion, we've decided to let queue messages be deleted regardless of whether there's an error in this step until we have real examples - they'll still be retried on the next polling interval, generally an hour later
- [x] Update ReadAndSend to not-retry any 400-range errors, not just 400 exactly. See https://docs.google.com/document/d/19xii5OlERxziIRfHftVzz8o1u9AHIbDdJlOb3B3mVVw/edit for expected errors
- [x] Refine the message we're sending to the queue
  - [x] Message to the queue should uniquely identify the partner
- [x] Add function deploy step to all environments
- [x] Continue evaluating options? https://www.notion.so/flexion-cdc-ti/Polling-external-sites-d2b6037e13ee463e933d5e475b106ffb
- [x] ~~Add Script to repo that will call both Docker and the requisite Azure Function code to stand everything up together in 1 command.~~
OR
- [x] Dockerize Azure Function
- [x] Extract cron into env var so that each environment is different so all environments don't hit at the same time 
- [x] Gather a list of outstanding questions for CA and get them sent out. 
- [x] TODO: Remove all TODO's from codebase by either doing them, making a separate task in a story, or just choosing to not do them.
- [x] Update function app to send logs to log analytics workspace

## Definition of Done
- [x] Documentation tasks completed
  - [x] Documentation and diagrams created or updated
    - [x] ADRs (`/adr` folder)
    - [x] Main `README.md`
    - [x] Other READMEs in the repo
  - [x] [Threat model](https://lucid.app/lucidchart/693ff5f9-c289-4585-9e74-a7f5abb36d69/edit?invitationId=inv_24d83a2b-b0fe-436d-8bd2-027b17beeb33) updated
- [x] Code quality tasks completed
  - [x] Code refactored for clarity and no design/technical debt
  - [x] Adhere to separation of concerns; code is not tightly coupled, especially to 3rd party dependencies
  - [x] Code is reviewed or developed by pair; 1 approval is needed but consider requiring an outside-the-pair reviewer
  - [x] Code quality checks passed
- [x] Security & Privacy tasks completed
  - [x] Security & privacy gates passed
- [x] Testing tasks completed
  - [x] Unit test coverage of our code >= 80%
- [x] Build & Deploy tasks completed
  - [x] Build process updated
  - [x] Source code is merged to the main branch

## Decisions
- *Optional: Any decisions we've made while working on this story*

## Notes
1.  Do the polling for California.

**Comment by halprin** [comment]
Thoughts on how to do this: https://www.notion.so/flexion-cdc-ti/Polling-external-sites-d2b6037e13ee463e933d5e475b106ffb?pvs=4

**Comment by jherrflexion** [comment]
We have a branch called Scheduling in the repository. Currently, we are exploring options with Terraform and GitHub Actions to establish a foundational state. We also hope to decouple the deployments between Ingestion and Polling. Once this is achieved, we plan to explore the options outlined in the Notion document.

**Comment by jherrflexion** [comment]
Currently we are adding environment variables needed to support the sftp service and polling. Once complete, we are testing the terraform and github actions to confirm we can deploy the polling.

**Comment by somesylvie** [comment]
Deploying another branch to `internal` destroyed and recreated the function app, so now the profile data is different than it was before and we're getting a 401 error trying to deploy the function app. Discussion in slack: https://flexion.slack.com/archives/C06CP7E3NAF/p1720739178841749?thread_ts=1720728627.874959&cid=C06CP7E3NAF

**Comment by jherrflexion** [comment]
Met with 3 Flexion Azure experts and unblocked ourselves with deploying the func.

**Comment by somesylvie** [comment]
The basic version of a scheduled Azure Function that enqueues a message (which we'll read in a later PR) is now working. Opened https://github.com/CDCgov/reportstream-sftp-ingestion/pull/78

**Comment by somesylvie** [comment]
Today we merged the PR mentioned in the comment above. Then we started refactoring the queue reader to allow space for the new queue and its data: https://github.com/CDCgov/reportstream-sftp-ingestion/pull/82
And updated the github workflows to deploy the new Azure Function in all envs instead of just internal: https://github.com/CDCgov/reportstream-sftp-ingestion/pull/84

**Comment by somesylvie** [comment]
Today we hooked up a first draft of the queue reader for the new polling queue, and it's mostly-working locally (we're having some kind of issue retrieving files with the local SFTP server, but I think that's not related to today's changes)

**Comment by somesylvie** [comment]
This PR hooks up the sftp work to the polling queue, sets some important settings, and makes it easier to run the Azure Function with all the other resources for this application https://github.com/CDCgov/reportstream-sftp-ingestion/pull/85

**Comment by jcrichlake** [comment]
This is in progress, we picked up this story since #1077 was blocked.  It's in progress, we have the functionality in place, we are working on adding tests and making sure the message settings are appropriate so that we don't lose data.

**Comment by pluckyswan** [comment]
The server public key has been acquired and is functional.

**Comment by somesylvie** [comment]
Currently waiting on the non-prod account from CADPH as well as answers to some more questions and the availability of testing data
